ibis/common/deferred.py

```diff
@@ -0,0 +1,556 @@
+from __future__ import annotations
+
+import collections.abc
+import functools
+import inspect
+import operator
+from abc import abstractmethod
+from typing import Any, Callable, TypeVar, overload
+
+from ibis.common.bases import FrozenSlotted as Slotted
+from ibis.common.bases import Hashable
+from ibis.common.collections import FrozenDict
+from ibis.util import is_iterable
+
+
+class Builder(Hashable):
+    """A deferred is a function that takes a context and returns a new object.
+
+    The context is a dictionary that contains all the captured values and
+    information relevant for the deferred. The deferred construct a new object
+    only given by the context.
+
+    The deferred is used in the right hand side of the replace pattern:
+    `Replace(pattern, deferred)`. Semantically when a match occurs for the
+    replace pattern, the deferred is called with the context and the result
+    of the deferred is used as the replacement value.
+    """
+
+    @abstractmethod
+    def build(self, context: dict):
+        """Construct a new object from the context.
+
+        Parameters
+        ----------
+        context
+            A dictionary containing all the captured values and information
+            relevant for the deferred.
+
+        Returns
+        -------
+        The constructed object.
+        """
+
+    @abstractmethod
+    def __eq__(self, other: Builder) -> bool:
+        ...
+
+
+class Deferred:
```

Comment:
The user facing wrapper on builder hierarchies, this is similar to `ibis.expr.types`. Since we need to overload `__eq__` the deferred objects should not be hashable (previously they had a hash implementation returning the id of the object to avoid collisions). 

On the other hand we need a hashable object structure for deferred objects to be used in the pattern system so I split the deferred base class into a different structure following the ibis IR abstraction.

Reply:
My intent here was to explicitly discourage the use of deferreds as dictionary keys. 

Imagine a use case like the following:

```py
# defined in module A
x = var("x")
mapping = {x: "a"}

# defined in module B
from moduleA import mapping

x = var("x")

mapping[x]  # raises KeyError
```

While it is logically correct, it is still surprising.






```

Currently they work uniformly because I restricted the `context` keys to be strings so `"_"` will be used across all deferred objects, but the previous implementation I had in patterns supported any kind of hashable keys for context. 

I was just more comfortable with

Reply:
That seems fair. I don't think we have any current APIs that support deferred values as dict keys anyway.